File: php/src/SSO/SingleSignOnRegistry.php
<?php

namespace SSO;

interface SingleSignOnRegistry
{
    /**
     * @param string $username
     * @param string $password
     * @return SSOToken
     */
    public function registerNewSession($username, $password);

    /**
     * @param SSOToken $token
     * @return boolean
     */
    public function isValid(SSOToken $token);

    /**
     * @param SSOToken $token
     */
    public function unregister(SSOToken $token);
}

File: cpp/src/SingleSignOnRegistry.cpp

#include <algorithm>

#include "SingleSignOnRegistry.h"

SSOToken *SingleSignOnRegistry::register_new_session(string &userName, string &password) {
    if (_gateway->credentialsAreValid(userName, password)) {
        auto *token = new SSOToken();
        _validTokens->push_back(token);
        return token;
    }
    return nullptr;
}

bool SingleSignOnRegistry::is_valid(SSOToken* token) {
    return std::find(_validTokens->begin(), _validTokens->end(), token) != _validTokens->end();
}

void SingleSignOnRegistry::unregister(SSOToken *token) {
    _validTokens->erase(std::remove(_validTokens->begin(), _validTokens->end(), token),
                        _validTokens->end());
}
